### PR TITLE
Manuel Cantu candidate-assignment-java

### PR DESCRIPTION
**Pull request by mcantureinhard**
### Short Description

Manuel Cantu

Models changed from interface to classes with @builder and @value
Relationships are made through references to the objects rather than a key
Additional comments in Comments.txt
No more optimizations to not deviate from original model

### Checks

- [ ] This PR follows the coding guidelines of the project
- [ ] This PR fulfills the acceptance criteria defined on the issue
- [ ] This PR has been locally tested
- [ ] Changes in this PR are unit tested

### CODE REVIEW

**Comment by mcantureinhard** [comment]
Actual commits:
commit c9e7e92d7f280d29832e49a68199109d8eb954e7
Author: Manuel Cantu Reinhard <m.cantu.reinhard@gmail.com>
Date:   Mon Jul 6 16:28:16 2020 +0200

    Add comments

commit 48a827bb112e10545dd25a85878627501fcbea59
Author: Manuel Cantu Reinhard <m.cantu.reinhard@gmail.com>
Date:   Mon Jul 6 16:07:27 2020 +0200

    Checkstyle fix

commit 3f1256e94eb6bbf747abaaa655bc9e02c444a5b3
Author: Manuel Cantu Reinhard <m.cantu.reinhard@gmail.com>
Date:   Mon Jul 6 15:50:04 2020 +0200

    Use object references, implement functions in Application

commit 3ae214ce56a9c3c91a7dec118b525996274adbc3
Author: Manuel Cantu Reinhard <m.cantu.reinhard@gmail.com>
Date:   Mon Jul 6 13:20:09 2020 +0200

    Use classes with builder and value instead of interfaces

File: src/main/java/ch/aaap/assignment/Application.java [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
 import ch.aaap.assignment.raw.CSVUtil;
 import java.time.LocalDate;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
+import java.util.stream.Collectors;
+import javafx.util.Pair;
```
Comment (robertblust) [inline src/main/java/ch/aaap/assignment/Application.java:19]: javafx? 🤔 cause of the usage of javafx.util.Pair, 

```
mvn clean install
```
fails. javafx is not declared as dependencies and I do not see why we need ui staff?

Comment (mcantureinhard) [reply src/main/java/ch/aaap/assignment/Application.java]: Hi Robert, apologies for the delay. My bad on that one. I just shift-enter the imports usually. I really thought there was something like std::pair. I just changed it to a list. Seems like that was the cause for the failure. It is strange that running the tests directly from intellij was working for me.

File: model.plantuml [hunk 2]
```diff
@@ -1,23 +1,26 @@
 @startuml
```
Comment (robertblust) [inline model.plantuml:1]: 🥇 for as well updating the plantuml file.

File: src/main/java/ch/aaap/assignment/Application.java [hunk 3]
```diff
@@ -24,10 +33,65 @@ private void initModel() {
...
+              .shortName(pc.getShortName())
+              .lastUpdate(pc.getLastUpdate())
+              .canton(cantons.stream().filter(c -> c.getCode().equals(pc.getCantonCode())).findFirst().orElse(null))
+              .district(districts.stream().filter(d -> d.getNumber().equals(pc.getDistrictNumber())).findFirst().orElse(null))
+              .build()
+      );
+    });
+
+    // Interesting fact that we can have postal communities that are part of multiple political communities.
+    // Had to change my model after I realized
```
Comment (robertblust) [inline src/main/java/ch/aaap/assignment/Application.java:61]: yes. such an easy topic Swiss Zip Codes get quite tricky once you deeper analyze the data 👍

File: src/main/java/ch/aaap/assignment/Application.java [hunk 4]
```diff
@@ -24,10 +33,65 @@ private void initModel() {
     Set<CSVPoliticalCommunity> politicalCommunities = CSVUtil.getPoliticalCommunities();
     Set<CSVPostalCommunity> postalCommunities = CSVUtil.getPostalCommunities();
 
-    // TODO implementation
-    throw new RuntimeException("Not yet implemented");
+    Set<District> districts = new HashSet<>();
+    Set<Canton> cantons = new HashSet<>();
+    Set<PoliticalCommunity> politicalCommunitiesSet = new HashSet<>();
+    Set<PostalCommunity> postalCommunitiesSet = new HashSet<>();
+    for (CSVPoliticalCommunity politicalCommunity : politicalCommunities) {
```
Comment (robertblust) [inline src/main/java/ch/aaap/assignment/Application.java:40]: I like the model initialization. Simple and pragmatic 👍

File: Comments.txt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+CSVs interesting facts:
+
+*Data not normalized over more files
+*One postal community can have multiple political communities
+*Based on the csv data loader KTKZ is the canton code, vs GDEKT. I was unable to verify elsewhere, but they should probably use the same name.
+(i.e. https://www.bfs.admin.ch/bfsstatic/dam/assets/214330/master )
+
+Other:
+Changed interfaces to classes to use @builder and @value which give us nice immutable objects.
```
Comment (robertblust) [inline Comments.txt:9]: I would do it the same. Interfaces makes sense when you expose something or have many different implementations 👍

File: Comments.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+First time using plantuml, unfortunately was not able get it to work with more than one relationship (syntax error).
+
+Since we are loading this data in memory I would use hashmaps instead of sets to store the data. Memory usage is the same, but we get a free index.
```
Comment (robertblust) [inline Comments.txt:13]: Good decision. Runtime performance is more important then initial performance 👍

File: Comments.txt [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+*Data not normalized over more files
+*One postal community can have multiple political communities
+*Based on the csv data loader KTKZ is the canton code, vs GDEKT. I was unable to verify elsewhere, but they should probably use the same name.
+(i.e. https://www.bfs.admin.ch/bfsstatic/dam/assets/214330/master )
+
+Other:
+Changed interfaces to classes to use @builder and @value which give us nice immutable objects.
+
+First time using plantuml, unfortunately was not able get it to work with more than one relationship (syntax error).
```
Comment (robertblust) [inline Comments.txt:11]: No sure what you mean here. 

This is the model created with plantuml from our reference solution
model

and here the model exposed as rest endpoint
https://platform-prod.3ap.ch/api/swagger/swagger-ui/index.html?configUrl=https%3A%2F%2Fplatform-prod.3ap.ch%2Fapi%2Fswagger%2Fconfig&urls.primaryName=zipcode-service

feel free to play around

**Pull request closed without merging**